sycl/plugins/opencl/pi_opencl.hpp

```diff
@@ -17,11 +17,99 @@
 #ifndef PI_OPENCL_HPP
 #define PI_OPENCL_HPP
 
+#include <regex>
+#include <string>
+
 // This version should be incremented for any change made to this file or its
 // corresponding .cpp file.
 #define _PI_OPENCL_PLUGIN_VERSION 1
 
 #define _PI_OPENCL_PLUGIN_VERSION_STRING                                       \
   _PI_PLUGIN_VERSION_STRING(_PI_OPENCL_PLUGIN_VERSION)
 
+namespace OCLV {
+class OpenCLVersion {
+protected:
+  unsigned int major;
+  unsigned int minor;
+
+public:
+  OpenCLVersion() : major(0), minor(0) {}
+
+  OpenCLVersion(unsigned int major, unsigned int minor)
+      : major(major), minor(minor) {
+    if (!isValid())
+      major = minor = 0;
+  }
+
+  OpenCLVersion(const char *version) : OpenCLVersion(std::string(version)) {}
+
+  OpenCLVersion(const std::string &version) : major(0), minor(0) {
+    /* The OpenCL specification defines the full version string as
+     * 'OpenCL<space><major_version.minor_version><space><platform-specific
+     * information>' for platforms and as
+     * 'OpenCL<space><major_version.minor_version><space><vendor-specific
+     * information>' for devices.
+     */
+    std::regex rx("OpenCL ([0-9]+)\\.([0-9]+)");
+    std::smatch match;
+
+    if (std::regex_search(version, match, rx) && (match.size() == 3)) {
+      major = strtoul(match[1].str().c_str(), nullptr, 10);
+      minor = strtoul(match[2].str().c_str(), nullptr, 10);
```

Comment:
From https://cplusplus.com/reference/cstdlib/strtoul/?kw=strtoul :
```
If the value read is out of the range of representable values by
an unsigned long int, the function returns ULONG_MAX (defined in 
[<climits>](https://cplusplus.com/climits)), and [errno]
(https://cplusplus.com/errno) is set to [ERANGE]
(https://cplusplus.com/ERANGE).
```
Suggest adding a check if the return value is `ULONG_MAX`, probably, the best place would be `isValid` function.